### ISSUE TITLE
what about the meaning of train_exist_nodes；num_prop；category_prop？

### ISSUE THREAD

**Issue by LaooBaoo** [Peien429/BotDGT#1]
抱歉，在阅读源代码尤其是在阅读Tainer类中的forward_one_batch方法时我认为自己没有正确理解这些变量的含义train_exist_nodes；num_prop；category_prop。
总的来说我有以下问题：
1.在动态图上，batch_size=64 的含义就是{Gt1,Gt2...Gtn}上最初选取该64个节点构图，在之后的时间片对应的图中**可能会加入其他非批次的节点**，还是说在之后的时间片中，依然是这64个节点之间建立或删除关系？因为我发现batch_id_n中的每个时间片上的维度可能会超过64所以有这个疑问。
2.train_exist_nodes的维度是时间片数*batch_size，但是我不理解所谓“exist”是什么含义？假如batch中选择的某一个节点在一个时间片上还并未与其他节点产生交互，则此时该节点的exist的值就为0吗？
3.num_prop和category_prop到底是什么含义？我完全没有头绪，他们的维度是torch.Size([229580, 5])和torch.Size([229580, 3])，这个“5”和“3”是怎么来的？

def forward_one_batch(self, batch_size, batch_n_id, batch_edge_index, batch_exist_nodes,
                          batch_clustering_coefficient, batch_bidirectional_links_ratio):
        des_tensor_list = [self.des_tensor[n_id].to(self.args.device) for n_id in batch_n_id]  
        tweet_tensor_list = [self.tweets_tensor[n_id].to(self.args.device) for n_id in batch_n_id] 
        num_prop_list = [self.num_prop[n_id].to(self.args.device) for n_id in batch_n_id]
        category_prop_list = [self.category_prop[n_id].to(self.args.device) for n_id in batch_n_id]
        label_list = [self.labels[n_id][:batch_size].to(self.args.device) for n_id in batch_n_id]
        label_list = torch.stack(label_list, dim=0)
        edge_index_list = [_.to(self.args.device) for _ in batch_edge_index]
        clustering_coefficient_list = [_.to(self.args.device) for _ in batch_clustering_coefficient]
        bidirectional_links_ratio_list = [_.to(self.args.device) for _ in batch_bidirectional_links_ratio]
        exist_nodes_list = [exist_nodes[:batch_size].to(self.args.device) for exist_nodes in batch_exist_nodes]
        exist_nodes_list = torch.stack(exist_nodes_list, dim=0)
        output = self.model(des_tensor_list, tweet_tensor_list, num_prop_list, category_prop_list, edge_index_list,
                               clustering_coefficient_list, bidirectional_links_ratio_list, exist_nodes_list,
                               batch_size)
        output = output.transpose(0, 1)
        loss = all_snapshots_loss(self.criterion, output, label_list, exist_nodes_list)
        return output, loss, label_list, exist_nodes_list
        
        希望您能帮我解决这些疑惑！

**Comment by Peien429** [comment]
您好，感谢您对我们的工作的关注。

1. 抱歉，这个代码上传比较仓促，我没有注意变量名称的可读性。
实际上，batch_n_id中的“batch”的含义并不是batch个节点，而是batch个snapshot；而且batch_n_id的batch与超参数batch_size没有关系，只与snapshot的数目有关系（再次抱歉，使用了batch这个字符串产生了歧义）。后面的 batch_edge_index, batch_exist_nodes, batch_clustering_coefficient, batch_bidirectional_links_ratio中的batch也是batch个snapshot的含义。
具体来说，以year为时间跨度为例，此时一共有13个snapshot。n_id代表1个snapshot下使用到的节点id，batch_n_id代表13个snapshot下的n_id构成的列表。在forward_one_batch中，首先根据13个snapshot的n_id，取出13个snapshot中的每个snapshot对应的节点特征并构成列表——des_tensor_list 、tweet_tensor_list 、num_prop_list和category_prop_list （这四个列表的长度也是13，代表13个snapshot下的节点特征）。其它基本同理，取出13个snapshot中的每个snapshot对应的edge_index、clustering_coefficient、bidirectional_links_ratio。label_list和exist_nodes_list 稍有不同，因为只需要对每个snapshot下的前batch_size个节点计算loss，所以截取了前batch_size个节点。

2. exist的含义是节点在某snapshot下是否存在，如果存在则为1，否则为0，具体逻辑可以在preprocess_twibot20.py中找到。使用exist的主要作用是，某个节点可能从第5个时间片才开始出现，那么该节点就不应该参与计算前4个snapshot下的loss，具体可以看loss.py文件。

3. num_prop和category_prop分别是Twibot-20和Twibot-22这两个benchmark中使用到的5个数字特征和3个类别特征，具体可以看Twibot-22的GitHub仓库。

**Comment by LaooBaoo** [comment]
> 您好，感谢您对我们的工作的关注。
> 
> 1. 抱歉，这个代码上传比较仓促，我没有注意变量名称的可读性。
>    实际上，batch_n_id中的“batch”的含义并不是batch个节点，而是batch个snapshot；而且batch_n_id的batch与超参数batch_size没有关系，只与snapshot的数目有关系（再次抱歉，使用了batch这个字符串产生了歧义）。后面的 batch_edge_index, batch_exist_nodes, batch_clustering_coefficient, batch_bidirectional_links_ratio中的batch也是batch个snapshot的含义。
>    具体来说，以year为时间跨度为例，此时一共有13个snapshot。n_id代表1个snapshot下使用到的节点id，batch_n_id代表13个snapshot下的n_id构成的列表。在forward_one_batch中，首先根据13个snapshot的n_id，取出13个snapshot中的每个snapshot对应的节点特征并构成列表——des_tensor_list 、tweet_tensor_list 、num_prop_list和category_prop_list （这四个列表的长度也是13，代表13个snapshot下的节点特征）。其它基本同理，取出13个snapshot中的每个snapshot对应的edge_index、clustering_coefficient、bidirectional_links_ratio。label_list和exist_nodes_list 稍有不同，因为只需要对每个snapshot下的前batch_size个节点计算loss，所以截取了前batch_size个节点。
> 2. exist的含义是节点在某snapshot下是否存在，如果存在则为1，否则为0，具体逻辑可以在preprocess_twibot20.py中找到。使用exist的主要作用是，某个节点可能从第5个时间片才开始出现，那么该节点就不应该参与计算前4个snapshot下的loss，具体可以看loss.py文件。
> 3. num_prop和category_prop分别是Twibot-20和Twibot-22这两个benchmark中使用到的5个数字特征和3个类别特征，具体可以看Twibot-22的GitHub仓库。

好的，谢谢！我差不多已经理解了。
